keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,667 @@
+
+There is a 1-to-many relationship between a `Bucket` and a `BucketRequest`, meaning that many `BucketRequest`s can reference the same `Bucket`.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  anonymousAccessMode: [6]
+    - private
+    - publicRead
+    - publicReadWrite
+  bucketClassName: [7]
+  bindings: [8]
+    - "<BucketAccess.name>"
+  protocol: [9]
+    protocolSignature: ""
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  allowedNamespaces: [10]
+    - name:
+      uid:
+  parameters: [11]
+status:
+  conditions: [12]
+    - Type: {BucketAvailable} [13]
+      Status: [14]
+      Reason: [15]
+      Message: [16]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"bucket-"<bucketRequest.name>"-"<bucketRequest.namespace>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.
+2. `labels`: added by the controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+3. `finalizers`: added by the controller to defer `Bucket` deletion until backend deletion ops succeed.
+4. `provisioner`: The provisioner field as defined in the `BucketClass`.  Used by sidecars to filter `Bucket`s. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+5. `releasePolicy`: Prescribes outcome of a Delete events. 
+   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
+   - _Delete_: the bucket and its contents are destroyed.
+> Note: the `Bucket`'s release policy is set to "Retain" as a default. Exercise caution when using the "Delete" release policy as the bucket content will be deleted.
+> Note: a `Bucket` is not deleted if it is bound to any `BucketRequest`s.
+6. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the storage instance or objects are intended to be publicly readable and/or writable.  Accepted values:
+   - `private`: Default, disallow uncredentialed access to the storage instance.
+   - `ro`: Read only, uncredentialed users can call ListBucket and GetObject.
```

Comment:
This seems not in line with the YAML above, which lists e.g., `publicReadWrite`, not `rw`.

Reply:
Good catch! I've changed this to a simple string with a single value and made it consistent with bucket class.